src/clusterfuzz/_internal/system/archive.py

```diff
@@ -34,6 +33,197 @@
     ZIP_FILE_EXTENSIONS + TAR_FILE_EXTENSIONS + LZMA_FILE_EXTENSIONS)
 
 
+def _is_attempting_path_traversal(archive_name, output_dir, filename) -> bool:
+  """_is_attempting_path_traversal"""
+  absolute_file_path = os.path.join(output_dir, os.path.normpath(filename))
+  real_file_path = os.path.realpath(absolute_file_path)
+
+  if real_file_path == output_dir:
+    # Workaround for https://bugs.python.org/issue28488.
+    # Ignore directories named '.'.
+    return False
+
+  if real_file_path != absolute_file_path:
+    logs.log_error('Directory traversal attempted while unpacking archive %s '
+                   '(file path=%s, actual file path=%s). Aborting.' %
+                   (archive_name, absolute_file_path, real_file_path))
+    return True
+  return False
+
+
+class ArchiveMemberInfo:
```

Comment:
Consider using a dataclass for this to automate the boilerplate:

```py
import dataclasses

@dataclasses.dataclass
class ArchiveMemberInfo:
  filename: str
  is_dir: bool
  file_size: int
  compress_size: int
```

Reply:
I didn't know about this, thanks, that's pretty cool!